core/src/main/java/org/apache/iceberg/encryption/KeyMetadata.java

```diff
@@ -0,0 +1,136 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.encryption;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+import static org.apache.iceberg.types.Types.NestedField.required;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.avro.AvroSchemaUtil;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.types.Types;
+
+class KeyMetadata implements EncryptionKeyMetadata, IndexedRecord {
+  private static final String encryptionKeyField = "encryption_key";
+  private static final String wrappingKeyIdField = "wrapping_key_id";
+  private static final String aadPrefixField = "aad_prefix";
+  private static final byte V1 = 1;
+  private static final Schema SCHEMA_V1 =
+      new Schema(
+          required(0, encryptionKeyField, Types.BinaryType.get()),
+          optional(1, wrappingKeyIdField, Types.StringType.get()),
+          optional(2, aadPrefixField, Types.BinaryType.get()));
+  private static final ThreadLocal<KeyMetadataEncoder<IndexedRecord>> ENCODER =
+      ThreadLocal.withInitial(() -> new KeyMetadataEncoder<>(V1));
+  private static final ThreadLocal<KeyMetadataDecoder<IndexedRecord>> DECODER =
+      ThreadLocal.withInitial(() -> new KeyMetadataDecoder<>(V1));
+
+  static final Map<Byte, Schema> supportedSchemaVersions = ImmutableMap.of(V1, SCHEMA_V1);
+
+  private String wrappingKeyId;
+  private ByteBuffer encryptionKey;
+  private ByteBuffer aadPrefix;
+
+  KeyMetadata() {}
+
+  KeyMetadata(ByteBuffer encryptionKey, String wrappingKeyId, ByteBuffer aadPrefix) {
+    this.wrappingKeyId = wrappingKeyId;
+    this.encryptionKey = encryptionKey;
+    this.aadPrefix = aadPrefix;
+  }
+
+  String wrappingKeyId() {
+    return wrappingKeyId;
+  }
+
+  ByteBuffer encryptionKey() {
+    return encryptionKey;
+  }
+
+  ByteBuffer aadPrefix() {
+    return aadPrefix;
+  }
+
+  static KeyMetadata parse(ByteBuffer buffer) {
+    KeyMetadataDecoder<IndexedRecord> decoder = DECODER.get();
+
+    try {
+      return (KeyMetadata) decoder.decode(buffer);
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to parse envelope encryption metadata", e);
+    }
+  }
+
+  @Override
+  public ByteBuffer buffer() {
+    KeyMetadataEncoder<IndexedRecord> encoder = ENCODER.get();
+
+    try {
+      return encoder.encode(this);
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to serialize envelope key metadata", e);
+    }
+  }
+
+  @Override
+  public EncryptionKeyMetadata copy() {
+    KeyMetadata metadata = new KeyMetadata(encryptionKey(), wrappingKeyId(), aadPrefix());
+    return metadata;
+  }
+
+  @Override
+  public void put(int i, Object v) {
```

Comment:
This is fine for now, but we will want to follow up after this PR with a fix for projection.